### PR TITLE
Feature/us 07

### PR DESCRIPTION
**Pull request by MukaddesUzdoganlar**
[UserStory_07](https://github.com/alhen002/capstone-event-repo/issues/13)

### CODE REVIEW

File: pages/index.js [hunk 1]
```diff
@@ -2,16 +2,56 @@ import useSWR from "swr";
...
+  function handleUpdateFilter(filterObject) {
+    setFilter((currentFilter) => {
+      return { ...currentFilter, ...filterObject };
+    });
+  }
+
+  const categories = events?.reduce(
+    (allCategories, curr) => [...allCategories, curr.category],
+    []
+  );
```
Comment (moonwave99) [inline pages/index.js:30]: Wouldn't this allow for duplicates if two events from the same category have the same?

A possible solution:

```js
function unique(array, key) {
  return [...new Set(array.map(x => x[key]))];
}

const events = [
  { id: 1, city: 'Berlin', category: 'a'},
  { id: 2, city: 'Berlin', category: 'a'},
  { id: 3, city: 'Hamburg', category: 'a'},
]

const cities = unique(events, 'city');
const categories = unique(events, 'category');
```

See the [Set docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)

File: pages/index.js [hunk 2]
```diff
@@ -2,16 +2,56 @@ import useSWR from "swr";
...
+
+  const {
+    data: events,
+    isLoading,
+    error,
+  } = useSWR(
+    `/api/events?queryCategory=${
+      filter?.category ? filter.category : ""
+    }&queryCity=${filter?.city ? filter.city : ""}`
+  );
```
Comment (moonwave99) [inline pages/index.js:19]: Too much ternary = nobueno

Abstract it in a function like:

```js
function getURL(filter = {}) {
  const params = new URLSearchParams();
  filter.category && params.append('category', filter.category);
  filter.city && params.append('city', filter.city);
  return `/api/events${params.size ? `?${params}` : ''}`;
}

console.log(getURL({ category: 'a' }));             // /api/events?category=a
console.log(getURL({ category: 'a', city: 'b' }));  // /api/events?category=a&city=b
console.log(getURL({ city: 'b' }));                 // /api/events?city=b
console.log(getURL());                              // /api/events
```

[See URLSearchParams docs](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) for more ^^

File: pages/api/events/index.js [hunk 3]
```diff
@@ -6,13 +6,19 @@ export default async function handler(request, response) {
 
   switch (request.method) {
     case "GET":
+      const { queryCategory, queryCity } = request.query;
```
Comment (moonwave99) [inline pages/api/events/index.js:9]: I would call them `category` and `city`, without the `query-` preamble ^^

File: pages/index.js [hunk 4]
```diff
@@ -2,16 +2,56 @@ import useSWR from "swr";
...
+
+  function handleUpdateFilter(filterObject) {
+    setFilter((currentFilter) => {
+      return { ...currentFilter, ...filterObject };
+    });
```
Comment (moonwave99) [inline pages/index.js:24]:
```js
setFilter((currentFilter) => ({ ...currentFilter, ...filterObject }));
```

File: pages/index.js [hunk 5]
```diff
@@ -2,16 +2,56 @@ import useSWR from "swr";
 import Error from "@/components/Error";
 import Loading from "@/components/Loading";
 import EventList from "@/components/EventList";
+import FilterBar from "@/components/FilterBar";
+import { useState } from "react";
 
 export default function HomePage() {
-  const { data: events, isLoading, error } = useSWR(`/api/events`);
+  const [filter, setFilter] = useState({});
```
Comment (moonwave99) [inline pages/index.js:9]: You can put the defaults here:

```js
useState({ city: '', category: '' });
```

**Review by moonwave99 — COMMENTED** [review]
Left some comments

**Review by moonwave99 — COMMENTED** [review]
Approved!

**Review by moonwave99 — APPROVED** [review]

**Comment by MukaddesUzdoganlar** [comment]
Hi Diego, thank you for your suggestions. We implemented all :-)

**Pull request merged**